shared-unpackaged/is-obscured.js

```diff
@@ -26,55 +34,70 @@ export default function isObscured(element: Element): boolean {
-    // checked them for visibility earlier.
-    const topLeftVisible =
-        topLeftElement &&
-        (element.contains(topLeftElement) || topLeftElement.contains(element));
-    const bottomRightVisible =
-        bottomRightElement &&
-        (element.contains(bottomRightElement) ||
-            bottomRightElement.contains(element));
-    return !topLeftVisible && !bottomRightVisible;
+    // TODO(somewhatabstract, WB-300): Need to cater to the case where the
+    // viewport is zoomed such that both corners are off screen but the rest
+    // isn't. In this case some browsers don't return the element from
+    // `elementFromPoint` then doesn't return the element.
+    // Also, consider how we might mitigate the pointer-events issue and make
+    // this call more robust.
+    const bounds = getBounds(anchorElement);
+
+    // This method does the main work, taking some coordinates and determining
+    // if our element is visible at that point or not.
+    const isVisible = (x: number, y: number): boolean => {
+        const elAtPoint = document.elementFromPoint(x, y);
+        if (
+            elAtPoint != null &&
+            (elAtPoint === popperElement || popperElement.contains(elAtPoint))
+        ) {
+            // Oh no, we're being obscured by our own popper.
+            // We need to look behind it. Shenanigans time.
+            const pointerEventsStyle = elAtPoint.style.pointerEvents;
+            // Remove pointer events so that we can look through it.
+            elAtPoint.style.pointerEvents = "none";
+            try {
+                const visible = isVisible(x, y);
+                return visible;
+            } finally {
+                // Make sure we put things back the way we found them. :)
+                elAtPoint.style.pointerEvents = pointerEventsStyle;
+            }
+        }
+        if (anchorElement instanceof Element) {
+            // If we are working with an element, then we can do some decendency checks
+            // to ensure we're not just hitting a child. We're ok with saying that
+            // we're visible if we hit a parent because we check them for visibility
+            // elsewhere.
+            return (
+                elAtPoint != null &&
+                (anchorElement.contains(elAtPoint) ||
+                    elAtPoint.contains(anchorElement))
+            );
+        }
+
+        // If element is a reference object, all we have to work with is
+        // intersection for checking obscurity. Since this doesn't cover
+        // parent/child relationships in the DOM, it's not really effective
+        // on its own and is possibly about as good as just returning `true`.
+        const intersection =
+            elAtPoint && getIntersection(bounds, getBounds(elAtPoint, true));
+        return (
+            intersection?.horizontal !== "within" ||
+            intersection?.vertical !== "within"
+        );
```

Comment:
Since adding flowtypes, I discovered that this function can get an `Element` or `ReferenceObject` and in the latter case, we seem to have much less to work with. I had noted, especially after hot-reloading, that we got errors around this problem, so I added some checks and code to cope with that. In truth, if we get here, our visibility check is much less useful, but at least it will do something half right now.